tests/rest/client/v1/test_login.py

```diff
@@ -748,3 +749,74 @@ def test_login_jwt_invalid_signature(self):
             channel.json_body["error"],
             "JWT validation failed: Signature verification failed",
         )
+
+
+AS_USER = "as_user_alice"
+
+
+class AppserviceLoginRestServletTestCase(unittest.HomeserverTestCase):
+    servlets = [
+        login.register_servlets,
+        register.register_servlets,
+        lambda hs, http_server: WhoamiRestServlet(hs).register(http_server),
+    ]
+
+    def register_as_user(self, username):
+        request, channel = self.make_request(
+            b"POST",
+            f"/_matrix/client/r0/register?access_token={self.service.token}",
```

Comment:
f strings are not supported on Python 3.5:

```suggestion
            "/_matrix/client/r0/register?access_token=%s" % (self.service.token,),
```